### ISSUE TITLE
Closure does not see that variable cannot be None

### ISSUE THREAD

**Issue by Gollum999** [python/mypy#9349]
* **Are you reporting a bug, or opening a feature request?**  Bug
* **Minimal example:**
```
$ mypy -c "
import typing as t                                                                                                  
def foo(x: t.Optional[t.Collection[int]] = None):
  if not x:
    x = []
  def bar() -> bool:
    return any(i > 0 for i in x)
"
```
* **What is the actual behavior/output?**
```
<string>:6: error: Item "None" of "Optional[Collection[int]]" has no attribute "__iter__" (not iterable)
Found 1 error in 1 file (checked 1 source file)
```
Changing the last line to `return any(x)` gives a slightly different error:
```
<string>:6: error: Argument 1 to "any" has incompatible type "Optional[Collection[int]]"; expected "Iterable[object]"
Found 1 error in 1 file (checked 1 source file)
```
* **What is the behavior/output you expect?**  No error.  This works as expected if a closure is not used.
* **What are the versions of mypy and Python you are using?**  Python 3.7.2, mypy 0.770
* **Do you see the same issue after installing mypy from Git master?**  I have not tried this yet
* **What are the mypy flags you are using? (For example --strict-optional)**
Here is my `.mypy.ini`:
```
[mypy]
# Specify the target platform details in config, so your developers are
# free to run mypy on Windows, Linux, or macOS and get consistent
# results.
python_version=3.7
platform=linux

# parse type info from imports, but don't emit warnings for them
follow_imports=skip

# treat Optional per PEP 484
strict_optional=True

# ensure all execution paths are returning
warn_no_return=True

# The following are off by default since they're too noisy.
# Flip them on if you feel adventurous.
disallow_untyped_defs=False
check_untyped_defs=False

[mypy-pandas.*]
# Type hints should be available in pandas v0.25
ignore_missing_imports=True

[mypy-numpy.*]
ignore_missing_imports=True

[mypy-statsmodels.*]
ignore_missing_imports=True

[mypy-patsy.*]
ignore_missing_imports=True
```

**Comment by hauntsaninja** [comment]
Duplicate of #2608, #9197, #9325 and others

**Comment by hauntsaninja** [comment]
tldr; things are tricky because the following will crash:

```
import typing as t                                                                                                  
def foo(x: t.Optional[t.Collection[int]] = None):
  if not x:
    x = []
  def bar() -> bool:
    return any(i > 0 for i in x)
  x = None
  bar()

foo()
```

**Comment by Gollum999** [comment]
Thanks for the example, I can totally appreciate how that would be difficult to fix.  I'll just stick with a `cast` for now.

Apologies for the duplicate.
